File: app/src/main/java/starbright/com/projectegg/dagger/module/StorageModule.java
/**
 * Created by Andreas on 5/5/2019.
 */

package starbright.com.projectegg.dagger.module;

import android.arch.persistence.room.Room;
import android.content.Context;
import android.content.SharedPreferences;
import android.preference.PreferenceManager;

import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;
import starbright.com.projectegg.dagger.qualifier.ApplicationContext;
import starbright.com.projectegg.data.local.RecipeDatabase;
import starbright.com.projectegg.util.Constants;

@Module
public class StorageModule {

    @Provides
    @Singleton
    SharedPreferences providesSharedPreferences(@ApplicationContext Context context) {
        return PreferenceManager.getDefaultSharedPreferences(context);
    }

    @Provides
    @Singleton
    RecipeDatabase provideDatabase(@ApplicationContext Context context) {
        return Room.databaseBuilder(context, RecipeDatabase.class,
                Constants.DATABASE_NAME).build();
    }
}

File: app/src/main/java/starbright/com/projectegg/dagger/module/StorageModule.kt
/*
 * Copyright (c) by Andreas (oentoro.andreas@gmail.com)
 * created at 5 - 9 - 2020.
 */

package starbright.com.projectegg.dagger.module

import android.content.Context
import android.content.SharedPreferences
import androidx.preference.PreferenceManager
import androidx.room.Room
import androidx.room.migration.Migration
import androidx.sqlite.db.SupportSQLiteDatabase
import dagger.Module
import dagger.Provides
import starbright.com.projectegg.dagger.qualifier.ApplicationContext
import starbright.com.projectegg.data.local.database.ApplicationDatabase
import javax.inject.Singleton

@Module
class StorageModule {

    @Provides
    @Singleton
    fun providesSharedPreferences(@ApplicationContext context: Context): SharedPreferences =
        PreferenceManager.getDefaultSharedPreferences(context)

    @Provides
    @Singleton
    fun getDatabase(@ApplicationContext context: Context): ApplicationDatabase =
        Room.databaseBuilder(
            context.applicationContext,
            ApplicationDatabase::class.java,
            "database"
        )
            .addMigrations(MIGRATION_1_2)
            .build()

    companion object {
        val MIGRATION_1_2 = object : Migration(1, 2) {
            override fun migrate(database: SupportSQLiteDatabase) {
                database.execSQL(
                    """
                    ALTER TABLE SearchHistory ADD COLUMN created_at INTEGER NOT NULL DEFAULT 0
                """.trimIndent()
                )
            }
        }
    }
}
